pkg/cache/v3/linear.go

```diff
@@ -42,13 +44,42 @@ type LinearCache struct {
 	watches map[string]watches
 	// Set of watches for all resources in the collection
 	watchAll watches
+	// Set of delta watches. A delta watch always contain the list of subscribed resources
+	// together with its current version
+	// version and versionPrefix fields are ignored for delta watches, because we always generate the resource version
+	deltaWatches map[int64]DeltaResponseWatch
+	// Continously incremented counter used to index delta watches
+	deltaWatchCount int64
+	// versionMap holds the current hash map of all resources in the cache.
+	// versionMap is only to be used with delta xDS.
+	versionMap map[string]string
 	// Continously incremented version
 	version uint64
 	// Version prefix to be sent to the clients
 	versionPrefix string
 	// Versions for each resource by name.
 	versionVector map[string]uint64
-	mu            sync.RWMutex
+
+	log log.Logger
+
+	mu sync.RWMutex
+}
+
+// This struct exist only because we want to reuse respondDelta function
+type cacheWrapper struct {
```

Comment:
I'm just curious as to what you think @s-matyukevich

Reply:
> We lose things like Update, Set, and Delete 

Not sure I follow this, why do we lose them? They are not present in the wrapper interface, but this interface is internal and should never be exposed. We still have them in the linear cache interface, and their implementation relies on cacheWrapper.

> it creates a rather large divergence with the linear SOTW implementation

Agree. But the difference in my opinion is that delta has a lot more things in common between simple and linear caches (It includes how we treat wildcard, calculate hashes, prepare RemovedResources list, and so on. Basically - all stuff we have in the respondDelta function)

>  How difficult would it be if we were to write linear from scratch with delta? I'd be willing to help, but it might require some heavy lifting. I'm inclined to keep the snapshot/resource wrapper stuff separated with the simple cache

I don't think it will be difficult and I considered this as an option. We could get rid of the cacheWrapper interface, copy respondDelta function to the linear.go file and modify it slightly. The only problem with this approach is that duplicates some logic from the respondDelta function (and in my opinion, this logic is similar/same for both delta and linear) If we duplicate this logic it might be harder to support in the future. If we decide to reuse this logic we must change the respondDelta function signature, as it currently accepts Snapshot, which is specific to the simple cache implementation (and that is exactly what I attempted to do)